core/modules/exact_match.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
+// Copyright (c) 2016-2017, Nefeli Networks, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_MODULES_EXACTMATCH_H_
+#define BESS_MODULES_EXACTMATCH_H_
+
+#include <rte_config.h>
+#include <rte_hash_crc.h>
+
+#include "../module.h"
+#include "../pb/module_msg.pb.h"
+#include "../utils/exact_match_table.h"
+
+using bess::utils::Error;
+using bess::utils::ExactMatchField;
+using bess::utils::ExactMatchKey;
+using bess::utils::ExactMatchRuleFields;
+using bess::utils::ExactMatchTable;
+using google::protobuf::RepeatedPtrField;
+
+typedef enum { FIELD_TYPE = 0, VALUE_TYPE } Type;
+
+class ExactMatch;
+class ValueTuple {
+  friend class ExactMatch;
+
+ public:
+  ValueTuple() : gate(), action() {}
+
+ private:
+  gate_idx_t gate;
+  ExactMatchKey action;
+};
```

Comment:
Is there any way to split this into base class of only gate and a derived class with action (or some other c++ oop concept). If it reduces the memory footprint and process batch performance impact, it would be good. ideally we match performance of upstream when there are no actions defined. Last I checked for simple case it was 50 vs 40 mpps.